mir.combinatorics: added binomial, cartesianPower, combinations, combinationsRepeat, permuations

wilzbach opened the pull request:
As discussed at https://github.com/D-Programming-Language/phobos/pull/4026#issuecomment-196493903 here comes `bionomial` - the first function for `mir`'s upcoming statistics package ;-)

I didn't know where to put it exactly - so I went ahead and created `stat.d` - feel free to suggest a better name for the module

greenify commented:
As requested at the [phobos PR](https://github.com/D-Programming-Language/phobos/pull/4026) here are `binomial`, `cartesianPower`, `combinations` and `combinationsRepeat`.

1) `bionomial` now supports arbitrary types like `bigint`
2) `bionomial` checks for overflows only if it's possible and returns zero in such a case
3) Logically `permutations` (from std.algorithms.iteration) felt to belong to this category, I copied it over as it wanted to modify it's documentation header and as if this is merged back to phobos will either go back into `iteration` or change `permutations` location.
4) there was a weird bug with dmd2.068 and ldc0.17 - `hasMember!(ulong, "max")` evaluates to false.
5) Complexities: I intend to go through the literature and find the according algortihms

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+    Binomial coefficient
+*/
+
+Result binomial(Result, NType, KType)(NType n, KType k)
```
`NType` should the same `KType`, just single `T` for both `n` и `k`.
`Result` should be replaced with `R = typeof(T.init - T.init)`.

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+}
+
+/// ditto
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
```
This prototype can be removed.

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+        static if (is(typeof((inout int = 0)
+        {
+            if (Result.max) {}
+        })))
```
`is(...) && !isFloatingPoint!R`

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+    assert(binomial(2, 1) == 2);
+}
+
+unittest
```
Needs test with floating point numbers (including fractional n and k)

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+                }
+
+                if (n % 2 == 1)
+                    swap(indices[0], indices[n-1]);
```
Please use `swapAt`

9il replied:
... in UFCS style `indices.swapAt(0, n-1);`

9il replied:
`indices.swapAt(0, h+1);`

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+                if (n % 2 == 1)
+                    swap(indices[0], indices[n-1]);
+                else
+                    swap(indices[state[n-2]], indices[n-1]);
```
ditto

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+                    return;
+                }
+
+                if (n % 2 == 1)
```
`if(n & 1)`

9il replied:
`if(h & 1)`

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+indexed) view into `r`.
+*/
+auto permutations(Range)(Range r)
+if (isRandomAccessRange!Range && hasLength!Range)
```
Needs an integration with `std.experimental.allocator`

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+{
+    static struct Permutations
+    {
+        size_t[] indices, state;
```
private?

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+
+        void popFront()
+        {
+            void next(int n)
```
Why this functions is needed?

9il replied:
Oh, I see. Please move this function out and make it private.

9il replied:
Better options: looks like this is just simple loop `for(size_t h = 0;;h++) { ... }` where `return` replaced with `break` and `h` is `n-2` (`h` would be a little bit faster).

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+module mir.combinatorics;
+
+private import std.range: isRandomAccessRange, hasLength;
```
import by default is private

9il replied:
change module to `std.range.primitives`

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+            _empty = r.length == 0;
+        }
+
+        private bool _empty;
```
Please move all variable declarations to the beginning of the struct

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+                // skip first duplicate
+                if (r.length > 1 && repeat > 1)
+                {
+                    _state = iota(repeat).array;
```
This is not good. First you allocate an array with `new`. Than you allocate another with `.array`. Please check other such cases.

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+                if (++state[n-2] == n)
+                {
+                    state[n-2] = 0;
+                    next(n+1);
```
just add `break` here

9il commented on source/mir/combinatorics.d:
```diff
@@ -0,0 +1,734 @@
+                {
+                    state[n-2] = 0;
+                    next(n+1);
+                }
```
```
if (state[h] == h+1)
{
    state[h] = 0;
    continue;
}
state[h]++;
break;
```

9il commented:
Something go wrong with LDC

greenify commented:
1) Wow - I just copied `permutations` over without looking deeply at it Phobos. I will remember to more critical to code from Phobos in the future.

2) swapAt is [not documentated](https://github.com/D-Programming-Language/phobos/blob/master/std/algorithm/mutation.d#L2004) - how about a PR at phobos?

3) I added support for negative number and floating point number

4) Floating point binomial runs into rounding issues for large n - e.g. [SciPy](https://github.com/scipy/scipy/blob/81c096001974f0b5efe29ec83b54f725cc681540/scipy/special/orthogonal_eval.pxd#L109) uses a more sophisticated algorithm. I will check if I find litrature about that, s.t. we can use something similar

5) Some addressed points in summary

> Result should be replaced with R = typeof(T.init - T.init).

ehm wouldn't this automatically result in the same type as T?
I prefer a different strategy - use the largest unsigned type (ulong) by default and let the user specify a different type if wished - this covers the following cases nicely

```D
// n and k are low numbers, the result is big
binomial!BigInt(1000, 10);
// n and k are by default int -> an overflow just happened
bionomial(100, 10);
```

> Needs test with floating point numbers (including fractional n and k)

what do you mean by this?
I added some tests for floating point numbers, but afaik there is no rational type yet in phobos

6) I added some constraints to the binomial coefficient, they might be a bit complicated, but what they basically say is that if there is a type that supports all these arithmetic operations, we can compute the binomial coefficient. The last line is there because bit comparison is only needed for signed types. The check for `isFloatingPoint` is optional (all floating types are already included because they support rudimentary arithemetic operations, however it makes the header easier to read.

7) ldc weirdness - probably this has something to do with the precision of real. I guess as soon as 4) is fixed, ldc will stop to complain.
ldc returns 4.47729e+85

8) will work on integration with the experimental allocator soon - do you like the way it is done for [EMSI containers](https://github.com/economicmodeling/containers)?

9il commented:
> 2) swapAt is not documentated - how about a PR at phobos?

Yes



> 5) Some addressed points in summary
> 
> Result should be replaced with R = typeof(T.init - T.init).
> ehm wouldn't this automatically result in the same type as T?

No, type of `byte - byte` is int. May be we ca use only `uint` and FP types?

greenify commented:
@9il I added an experimental interface to the allocator API.
It does support adding a custom allocator now and even `@nogc` works :)

- it stops being `pure` (that's why I still use `new` for the default GCAllocator)
- it expects the allocator type to have a `.instance` property - I am not sure whether this is a fair assumption, but it simplifies our code a lot.. 
- only if version is `> 2.069` (hence a few conditionals in the code).

What do you think about this design?

> No, type of byte - byte is int. May be we ca use only uint and FP types?

Negative binomial is defined too. What is your opinion about my current solution with the `isArithmetic` template and having `ulong` as default target type that can be overwritten if needed?

9il commented:
Please drop support for dmd FE <= 2.068 for this module.

-----
No all allocators has `instance`. Not all are shared.

```D
(Allocator, Range)(Range r, Allocator alloc)
```

and wrapper for `GCAllocator`:
```D
(Range)(Range r)
{
// import GCAllocator internal
}
```

9il commented:
> isArithmetic

Good idea to reduce imports, thanks!

greenify commented:
> Please drop support for dmd FE <= 2.068 for this module.

I would like to, but the allocator was introduced in 2.070, so even LDC would fail.
Btw that's also why I didn't use local GCAllocator import's yet, because we still need the stub.

> No all allocators has instance.

I updated `combinationsRepeat`, making Allocator optional has a couple of caveats:
- we have to use two methods instead of one (that's why I made the struct public)
- we have to have two different constructors (as GCAllocator.instance is already impure - we cannot access mutable static data 'instance')

> Not all are shared.

At least GCAllocator needs to be `shared` - is there any way we can infer this?

See `combinationsRepeat`:

https://github.com/DlangScience/mir/pull/29/files#diff-def635b35c404dc25be3ab3c785e833fR778

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+}
+body
+{
+
```
blankline

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+
+    static if(!is(Allocator == GCAllocator))
+    {
+        this(Range r, size_t repeat, shared Allocator alloc)
```
remove `shared`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+}
+body
+{
+    return CombinationRepeat!(GCAllocator, Range)(r, k);
```
need import for `GCAllocator`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+
+static if (__VERSION__ > 2069)
+{
+    import std.experimental.allocator.gc_allocator: GCAllocator;
```
should be removed from global

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+            }
+        }
+
+        static if(!is(Allocator == GCAllocator))
```
~~Probably this can solved by user~~

9il replied:
hmm, lets use this code, thought

9il commented:
> At least GCAllocator needs to be shared - is there any way we can infer this?

No worries. `shared` would be a part of type in template difinition

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+
+            size_t rLength = r.length;
+
+            static if(is(Allocator == GCAllocator))
```
This type of conditions can be removed, just use common case :)

wilzbach replied:
I would like too, but then we loose the purity of the struct and function.

9il replied:
OK, lets keep the code

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1006 @@
+
+        this(Range r, shared Allocator alloc)
+        {
+            import std.range : iota;
```
Try to avoid imports if possible

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,987 @@
+    // big floating point
+    assert(binomial!real(1.2e10, 2.3).approxEqual(5.66794e22, 0.0001));
+    // TODO: rounding errors seem to sum up here
+    //assert(binomial!real(2.245e15, 5.7).approxEqual(7.68342e84, 0.1));
```
this might be a future problem, when using binomial with floats/reals

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,987 @@
+indexed) view into `r`.
+*/
+auto permutations(Allocator = bool, Range)(Range r,
+                  Allocator alloc = Allocator.init)
```
using `bool` as default type is a hack, s.t. that we have just one constructor and function. It makes the constructor easier. Is there a better default type that can be used a parameter?

The alternative is to put the struct out as `private struct` as done before.

9il replied:
Oh, don't do this kind of hack please)

wilzbach replied:
You prefer a `private struct` or is there a better way?

9il replied:
Why you don't want to have two prototypes ?

9il replied:
private struct is OK

wilzbach replied:
because there is still some shared part of the constructor - ok will move that part into a own function and put everything to the struct ;-)
Thanks for your input!

9il replied:
> because there is still some shared

???

9il replied:
Ah, understand now)

wilzbach replied:
already fixed - don't know why github is still showing this.

On 2016-04-09 17:32, Ilya Yaroshenko wrote:

> > -    //assert(binomial!real(2.245e15, 5.7).approxEqual(7.68342e84, 0.1));
> >   +}
> >   +
> >   +/**
> >   +Lazily computes all _permutations of `r` using $(WEB
> >   +en.wikipedia.org/wiki/Heap%27s_algorithm, Heap's algorithm).
> >   +
> >   +While generating a new item is in `O(k)` (amortized `O(1)`),
> >   +the number of permutations is `|r|!`.
> >   +
> >   +Returns:
> >   +A forward range the elements of which are an $(XREF range,
> >   +indexed) view into `r`.
> >   +*/
> >   +auto permutations(Allocator = bool, Range)(Range r,
> > -                  Allocator alloc = Allocator.init)
> > 
> > because there is still some shared
> 
> ???
> 
> ---
> 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/DlangScience/mir/pull/29/files/af37048c8931dea6638ff5e4d59ff5ed89e801f8#r59116036

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1089 @@
+A forward range the elements of which are an $(XREF range,
+indexed) view into `r`.
+*/
+auto permutations(Range, Allocator)(Range r, auto ref Allocator alloc)
```
Inspired by your allocate API PR:
- I changed the order, so that the type can automatically be found
- `alloc` is passed as `ref` and only stored as a pointer

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1089 @@
+    private void _init()
+    {
+        _empty = r.length == 0;
+    }
```
I could put this into both constructor, but for other structs (below) it's more logic - so I went for a common behavior.

9il replied:
???

9il replied:
default value would be 0

wilzbach replied:
The other constructors share some common bits - I put them into such a init function to avoid duplication.

9il replied:
Oh, my bad. Have not seen `==`

wilzbach commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1089 @@
+    return Permutations!Range(r);
+}
+
+private struct Permutations(Range, Allocator = void)
```
I needed a way to pass the Allocator optionally to the struct - as GCAllocator is not currently available, I used `void`. This time it is _hidden_ from the user and I actually like it a lot more than the `GCAllocator` stub.

9il replied:
`private` should be removed

9il replied:
`Allocator` should be removed too

9il replied:
just construct this struct inside a function

wilzbach replied:
How? I have to save the reference to the Allocator somehow and afaik it isn't possible to extend a struct.

9il replied:
add a `dispose` method for user

wilzbach replied:
Okay that looks nicer - thanks!

```
    this(Allocator)(Range r, ref Allocator alloc)
    {
        ...
    }

    void dispose(Allocator)(Allocator alloc)
    {
       ...
    }
```

But isn't it pretty inconvenient (and easy to forgot) for an user to dispose the structs internal state values manually?
AFAICT as soon as the structs gets destructed, they are not needed anymore. 

Moreover the other `structs` support `save`, which also needs allocation, thus saving the reference to the allocation is needed imho.

wilzbach replied:
Update: Compare [permutations](https://github.com/DlangScience/mir/pull/29/files#diff-def635b35c404dc25be3ab3c785e833fR203) (not saving the reference to the allocator) with [cartesianPower](https://github.com/DlangScience/mir/pull/29/files#diff-def635b35c404dc25be3ab3c785e833fR403).
The current permutation test also doesn't call dispose (yet).

wilzbach replied:
Another point for storing the Allocator ref:

> It is assumed the respective entities had been allocated with the same allocator.

(from the [dispose api docs](https://dlang.org/library/std/experimental/allocator/dispose.html))

wilzbach commented:
@9il - I reworked the allocation support and moved all nested the structs out.
This time I am actually happy with the result. Thanks you so much for your helpful & enormous feedback that help{ed,s} me to learn.

As this is getting in its final shape - I squashed all commits into one.

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1090 @@
+        _empty = r.length == 0;
+    }
+
+    this(Range r)
```
should accept indexes and state arrays

wilzbach replied:
the other structs support the typical `save` (and now permutations too) - is this what you are asking for?

9il replied:
``` D
/// UPDATED
this(Range r, size_t[] indices, size_t[] state)
in{
    size_t rLength = r.length;
    assert(state.length = rLength - 1);
    assert(indices.length = rLength);
}
body {
    foreach(i, ref index; indices)
        index = i; //less imports
    state[] = 0;
    this.indices = indices;
    this.state = state;
    this.r = r;
}
```

9il replied:
`indices` and `state` should be allocated in `permutations` / `makePermutations`

wilzbach replied:
I am afraid I don't understand why you would expose these two arrays to the user?
Those are internal state variables that are likely to change. A future version of permutations might only need `state` or use more allocated variables.

An alternative would be to use an allocated `struct` which contains this data like

```
PermutationsAllocation permsAllocated = makePermutations(Allocator alloc, 4); // we need k for the length
auto perms = permutations(4, permsAllocated); // weneed k again
dispose(permsAllocated, alloc);
```

In any case I also don't understand why you want to force the user to call `makePermutations` and `dispose` himself. Do you mind to explain this?

The most popular package that I know of that uses the allcoator API are the gcfree [containers](https://github.com/economicmodeling/containers). They save the allocator too.

9il replied:
> In any case I also don't understand why you want to force the user to call makePermutations and dispose himself. Do you mind to explain this?

I don't force user to do something, I just want to have more control. Anyway a user will be able to use GC version without any additional thoughts.

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1090 @@
+            _init();
+        }
+
+        ~this()
```
just remove this

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1090 @@
+        _init();
+    }
+
+    static if(!useGC)
```
move to the `make ***` function

wilzbach replied:
Do you mean I should add a special `make` function for allocation?

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1136 @@
+        _init();
+    }
+
+    this(Allocator)(Range r, ref Allocator alloc)
```
please move this logic to `makePermutations`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1136 @@
+        _init();
+    }
+
+    void dispose(Allocator)(Allocator alloc)
```
please move this outside of the struct

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1136 @@
+        typeof(this) c = this;
+        size_t rLength = indices.length;
+
+        c.indices = alloc.makeArray!size_t(rLength);
```
`c.indices = alloc.makeArray!size_t(indices);`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1136 @@
+        foreach(i, r; indices)
+            c.indices[i] = indices[i];
+
+        c.state = alloc.makeArray!size_t(rLength - 1);
```
`c.state = alloc.makeArray!size_t(state);`

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1136 @@
+        return c;
+    }
+
+    @property typeof(this) save(Allocator)(ref Allocator alloc)
```
move this outside the struct

wilzbach replied:
`save` is the standard method in Phobos to save the state of a struct / function. Imho it should work when using an Allocator too - otherwise a lot of the range algorithms will not work. Most of them expect a `ForwardRange`.
https://dlang.org/library/std/range/interfaces/forward_range.html

9il replied:
i asked only about `save` with allocator. Ok lets preserve both in struct

wilzbach replied:
Ok I guess this is really about whether we should keep a reference (or copy) to the allocator in our struct or not. If we keep the reference.

`save` is called without any variable in those algorithms, hence I would make `static if` variants of `save` depending on whether an Allocator was passed - otherwise I guess moving `save` out makes sense.

9il replied:
Algorithms can be extended with save with allocators ;)

9il replied:
> Algorithms can be extended with save with allocators ;)

hmm, this was bad idea

9il commented:
After thinking about API and algorithms I would like to suggest to remove all templated code (except `make***` functions) and use Ranges of indexes instead of templates.
This allows:
1. Reduce code size
2. Allows user to use same indexes with different ranges without ugly `zip` magic
3. This will simplified code

In the same time user will be able to achieve the same result as with template using `perm.map!(p => userRange.indexed(p))`.

9il commented:
> After thinking about API and algorithms I would like to suggest to remove all templated code (except make*** functions) and use Ranges of indexes instead of templates.
> This allows:
> 1. Reduce code size
> 2. Allows user to use same indexes with different ranges without ugly zip magic
> 3. Maintain more simple code
> 
> In the same time user will be able to achieve the same result as with template using perm.map!(p => userRange.indexed(p)).

What do you think about?

wilzbach commented:
> What do you think?

I am currently traveling thus the delayed answer,but afaik one still has to do the allocation, do you don't gain anything, but make it more complex.

9il commented:
> I am currently traveling thus the delayed answer,but afaik one still has to do the allocation, do you don't gain anything, but make it more complex.
>  Show all checks

Allocation should be done in `permutations` and `makePermutations`. Do you see any problems with it?

9il commented:
BTW, would you mind me merge this and open PR with changes?

wilzbach commented:
> BTW, would you mind me to merge this and open PR with changes?

Not at all. Should I squash the commits?

9il commented:
> Should I squash the commits?

When this would be ready.

So, you don't want to separate logic and templates?

wilzbach commented:
> So, you don't want to separate logic and templates?

My first choice would be to use just the allocation api, but this leads to an impure struct, but yeah I do prefer having a convenient api for the user and this means to combine the logic and templates.

9il commented:
> combine the logic and templates

because you want to store allocator's pointer?

My point about convenient api is simple: if user wants convenient and simple API he will use GC. If user wants to use allocators he need to think about allocation logic.

9il commented:
Note, that this is not container. This simple a permutation range

wilzbach commented:
> Note, that this is not a container. This is a simple permutation range

Exactly my point that I am trying to make. In 99% of the use cases a user wants just to iterate through the range and once he his finished and leaves the scope, he doesn't need it anymore. Thus  automatic deallocation should happen. 

Btw I would be very nice of the language would have more built in support for the allocation api..

9il commented:
> In 99% of the use cases

... user will use GC.

9il commented:
Allocators support is for libraries developers, not for a common user. And libraries developers needs full control with minimal effort.

wilzbach commented:
Okay - do you want to post a suggestion how you imagine the api to look like? I guess we are a lot faster this way.

9il commented:
```D
Permutations permutations(size_t n) @safe // and etc
{
	assert(n); //???
	return Permutations(new state[n-1], new indices[n-1]);
}

unittest
{
	import std.algorithm.iteration: indexed; //?
	auto rp = permutations(r.length).map!(perm => r.indexed(perm));
	//...
}

Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
{
	import std.experimental.allocator: makeArray;
	//...
}


void dispose(Allocator)(auto ref Allocator alloc, auto ref Permutations perm)
{
	import std.experimental.allocator: dispose;
	//...
}

struct Permutations //Just a struct
{
	this(size_t[] indices, size_t[] state) @safe // and etc
	in{
		assert(state.length + 1 = indices.length);
	}
	body {
		foreach(i, ref index; indices)
			index = i; //less imports
		state[] = 0;
		this.indices = indices;
		this.state = state;
	}
}
```

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -947,7 +945,7 @@ public:
 
         // behaves like: do _getNextState();  while(!_state.isSorted);
-        size_t i = repeat - 1;
+        uint i = cast(uint) repeat - 1;
```
why `i` is uint?

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -961,7 +959,7 @@ public:
         if (i != repeat - 1)
         {
-            for (size_t j = i + 1; j < repeat; j++)
+            for (uint j = i + 1; j < repeat; j++)
```
why `j` is uint? why not simple foreach loop?

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -915,13 +913,13 @@ private struct CombinationsRepeat
 
 public:
-    this(size_t n, size_t[] state) @safe pure nothrow @nogc
+    this(uint n, uint[] state) @safe pure nothrow @nogc
```
use `size_t` here

9il replied:
.. for `n`

wilzbach replied:
why? - it will be called via the wrappers and they have `cast` k too

9il replied:
ditto

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -915,13 +913,13 @@ private struct CombinationsRepeat
-    this(size_t n, size_t[] state) @safe pure nothrow @nogc
+    this(uint n, uint[] state) @safe pure nothrow @nogc
     {
         this.n = n;
```
but cast here

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -254,10 +254,10 @@ Returns:
 indexed) view into `r`.
 */
-Permutations permutations(size_t n) @safe pure nothrow
+Permutations permutations(uint n) @safe pure nothrow
```
~~make cast inside the functiton~~

9il replied:
Why you use uint anywhere????
We need it only for internal data storage !

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -301,12 +301,12 @@ pure @safe nothrow unittest
 
 /// ditto
-Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
+Permutations makePermutations(Allocator)(auto ref Allocator alloc, uint n)
```
ditto
`size_t n`

wilzbach commented:
> make cast inside the functiton

[done](https://github.com/DlangScience/mir/pull/29/commits/9c076b608bcb5a8a1e51abbfce52463289004630)

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -257,7 +257,8 @@ indexed) view into `r`.
 {
     assert(n, "must have at least one item");
-    return Permutations(new size_t[n-1], new size_t[n]);
+    uint nu = cast(uint) n;
```
Why do you need cast here???????????

wilzbach replied:
because `uint`[size_t n]` doesn't work

9il replied:
it works

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -427,7 +427,7 @@ in
 body
 {
-    return CartesianPower(n, new size_t[repeat]);
+    return CartesianPower(cast(uint) n, new uint[cast(uint) repeat]);
```
please move cast to the constructor

wilzbach replied:
I need to cast `repeat` - as uint[size_t repeat]` is invalid - therefore casting at the same place made sense to me

9il replied:
> uint[size_t repeat]` is invalid

this is not true

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -451,7 +451,7 @@ in
 {
     import std.experimental.allocator: makeArray;
-    return CartesianPower(n, alloc.makeArray!size_t(repeat));
+    return CartesianPower(cast(uint) n, alloc.makeArray!uint(cast(uint) repeat));
```
ditto

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -613,11 +613,11 @@ in
 body
 {
-    return Combinations(n, new size_t[k]);
+    return Combinations(cast(uint) n, new uint[cast(uint) k]);
```
ditto

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -651,17 +651,17 @@ private struct Combinations
 public:
 
-    this(size_t n, size_t[] state) @safe pure nothrow @nogc
+    this(uint n, uint[] state) @safe pure nothrow @nogc
```
ditto

wilzbach commented:
> > uint[size_t repeat]` is invalid
> this is not true

Oh I am terribly sorry - that was my naive understanding then :<
I am impressed how much I can do wrong. Thanks for your patience!

wilzbach commented:
![image](https://cloud.githubusercontent.com/assets/4370550/14429116/be482754-0004-11e6-9e82-0c4bdf30819e.png)

Seems like `/// ditto` isn't working that well?

9il commented:
> Seems like /// ditto isn't working that well?

ddox is not very well

wilzbach commented:
well I guess we have to switch to a better documentation engine soon :/

![image](https://cloud.githubusercontent.com/assets/4370550/14431029/79ffb1f4-000c-11e6-961c-8a4a47f33e99.png)

btw what do you think about naming the module `comb` or `combs`?
(I like short names)

9il commented:
> btw what do you think about naming the module comb or combs?

`combinatorics` looks clear, imho

wilzbach commented:
> combinatorics looks clear, imho

Okay :)

Anything else you want to see in the initial version?

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1109 @@
+    if ((isArithmetic!(R, T) || isFloatingPoint!R) &&
+        ((is(typeof(T.min < 0)) && is(typeof(T.init & 1))) || !is(typeof(T.min < 0))) )
+{
+    import std.traits: isFloatingPoint;
```
can be removed

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1109 @@
+    assert(a.front.equal([0, 1]));
+}
+
+/// ditto
```
This should be documented

wilzbach replied:
does `void` need a return type?

9il replied:
no

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1109 @@
+        _empty = indices.length == 0;
+    }
+
+    @property uint[] front() @safe pure nothrow @nogc
```
Document with empty `///`

9il replied:
and other range primitives

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1109 @@
+/**
+Creates a projection of `collection` Range (e.g. `Combinations`)
+onto a custom `range`.
+*/
```
Params and Returns for all public functions

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1109 @@
+    alloc = custom Allocator
+
+Returns:
+    Forward range which yields the permutations
```
`Forward range, which ...` if i am not wrong

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1109 @@
+    dispose(alloc, cartesianPower._state);
+}
+
+private struct CartesianPower
```
BTW, this struct should be private/package. Just undocumented. Otherwise user would not be able to access methods

wilzbach commented:
Comments addressed in d63ff57.

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1150 @@
+    dispose(Mallocator.instance, rp);
+}
+
+/// ditto
```
What /// ditto?

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1150 @@
+    }
+
+    ///
+    @property uint[] front() @safe pure nothrow @nogc
```
`const(uint)[]` and for other structs too.
Also the note about that this range would be changed by the struct after `popFront`, for other structs too

wilzbach replied:
> Also the note about that this range would be changed by the struct after popFront, for other structs too

What do you mean by that?

wilzbach replied:
> const(uint)[] and for other structs too.

addressed.

9il replied:
> Also the note about that this range would be changed by the struct after popFront, for other structs too
> What do you mean by that?

Oh, my English ... :/

We need to note a user that popFront will change content of the returned range

wilzbach replied:
> Oh, my English ... :/

No worries - that's how i understood it.

> We need to note a user that popFront will change content of the returned range

Isn't that expected for ranges and popFront?
That's why there are forward ranges that one can save?

9il replied:
```
const(uint)[] data = p.front;
p.popFront; // `data` is changed here
```

wilzbach replied:
I see - thanks :)
But modifying the indexes doesn't make sense, so why don't we make it entirely `const`?

9il replied:
> But modifying the indexes doesn't make sense, so why don't we make it entirely const?

To disallow a user to break our indices

wilzbach replied:
> To disallow a user to break our indices

Oh I was proposing `const uint[]` - as far as i understand this implies constness for its values.

9il commented on source/mir/combinatorics/package.d:
```diff
@@ -0,0 +1,1149 @@
+    }
+
+    ///
+    @property const(uint[]) front() @safe pure nothrow @nogc
```
`const(uint)[]`, but not const(uint[])

wilzbach replied:
sorry - typo :/

wilzbach commented:
no rebase?

9il commented:
O_o

wilzbach commented:
> O_o

I am sorry - I tried to avoid total rebases in the end, so that reviewing the changes is easier for you.
Do you want to do a `git rebase` and push force?

9il commented:
> I am sorry - I tried to avoid total rebases in the end, so that reviewing the changes is easier for you.
> Do you want to do a git rebase and push force?

No, they are squashed

wilzbach commented:
Ah wow - my fault. Didn't see that.

Btw about the `const` discussion - how about changing `const(uint)[] front()` to `const uint[] front()`?
(Then the user can't change neither the list nor the elements)

9il commented:
this is bad api
> On 11 апр. 2016 г., at 19:19, Sebastian Wilzbach <notifications@github.com> wrote:
> 
> Ah wow - my fault. Didn't see that.
> 
> Btw about the const discussion - how about changing const(uint)[] front() to const uint[] front()?
> (Then the user can't change neither the list nor the elements)
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub <https://github.com/DlangScience/mir/pull/29#issuecomment-208456524>

wilzbach commented:
> this is bad api

why? didn't you want to add as a notice that the user should be careful when modifying the array itself?

9il commented:
He can change array but not content

On Mon, Apr 11, 2016, 19:23 Sebastian Wilzbach <notifications@github.com>
wrote:

> this is bad api
>
> why? didn't you want to add as a notice that the user should be careful
> when modifying the array itself?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/DlangScience/mir/pull/29#issuecomment-208458281>
>

wilzbach commented:
> He can change array but not content

Yep but it is the internal state array, so changing it will lead most likely to an error. 
For a normal user it is proxied via the IndexedRoR struct, so he won't be affected by this constraint?

9il commented:
Yes

> On 11 апр. 2016 г., at 19:31, Sebastian Wilzbach <notifications@github.com> wrote:
> 
> He can change array but not content
> 
> Yep but it is the internal state array, so changing it will lead most likely to an error. 
> For a normal user it is proxied via the IndexedRoR struct, so he won't be affected by this constraint?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub <https://github.com/DlangScience/mir/pull/29#issuecomment-208464700>

wilzbach commented:
> Yes

So why is using `const uint[]` a bad API then? Sorry that I still don't understand your reasoning.

9il commented:
> So why is using const uint[] a bad API then? Sorry that I still don't understand your reasoning.

```D
auto p = permutations(4);
auto p = p.front;
p = new uint[3];// fails if `const uint[]`
```

wilzbach commented:
Thanks for the example and sorry for bothering you!
